webapp/tests/karma/ts/services/message-contact.spec.ts

```diff
@@ -0,0 +1,292 @@
+      ]));
+
+      return service
+        .getConversation('abc')
+        .then(result => {
+          expect(query.args[0][1]).to.deep.equal(expectedQueryParams);
+          expect(getDataRecordsService.get.callCount).to.equal(0);
+          expect(hydrateMessagesService.hydrate.callCount).to.equal(1);
+          expect(hydrateMessagesService.hydrate.args[0]).to.deep.equal([[
+            { id: 'some_id1', value: { id: 'id1' }, doc: { _id: 'some_id1' } },
+            { id: 'some_id2', value: { id: 'id2' }, doc: { _id: 'some_id2' } },
+            { id: 'some_id3', value: { id: 'id3' }, doc: { _id: 'some_id3' } },
+            { id: 'some_id4', value: { id: 'id4' }, doc: { _id: 'some_id4' } },
+          ]]);
+          expect(result).to.deep.equal([
+            { id: 'some_id1', value: { id: 'id1' }, doc: { _id: 'some_id1' }, hydrated: true, read: true },
+            { id: 'some_id2', value: { id: 'id2' }, doc: { _id: 'some_id2' }, hydrated: true, read: true },
+            { id: 'some_id3', value: { id: 'id3' }, doc: { _id: 'some_id3' }, hydrated: true, read: true },
+            { id: 'some_id4', value: { id: 'id4' }, doc: { _id: 'some_id4' }, hydrated: true, read: true },
+          ]);
+        });
+    });
+
+    it('should build conversation with skip', () => {
+      const expectedQueryParams = {
+        reduce: false,
+        descending: true,
+        include_docs: true,
+        skip: 45,
+        limit: 50,
+        startkey: [ 'abc', {} ],
+        endkey: [ 'abc' ]
+      };
+      const query = sinon.stub().returns(Promise.resolve({}));
+      dbService.get = () => ({
+        query,
+        allDocs: sinon.stub().returns(Promise.resolve({}))
+      });
+      hydrateMessagesService.hydrate.returns(Promise.resolve([]));
+
+      return service
+        .getConversation('abc', 45)
+        .then(result => {
+          expect(query.args[0][1]).to.deep.equal(expectedQueryParams);
+          expect(getDataRecordsService.get.callCount).to.deep.equal(0);
+          expect(hydrateMessagesService.hydrate.callCount).to.equal(1);
+          expect(hydrateMessagesService.hydrate.args[0]).to.deep.equal([[]]);
+          expect(result).to.deep.equal([]);
+        });
+    });
+
+    it('should build conversation with limit', () => {
+      const expectedQueryParams = {
+        reduce: false,
+        descending: true,
+        include_docs: true,
+        skip: 45,
+        limit: 120,
+        startkey: [ 'abc', {} ],
+        endkey: [ 'abc' ]
+      };
+      const query = sinon.stub().returns(Promise.resolve({}));
+      dbService.get = () => ({
+        query,
+        allDocs: sinon.stub().returns(Promise.resolve({}))
+      });
+      hydrateMessagesService.hydrate.returns(Promise.resolve([]));
+
+      return service
+        .getConversation('abc', 45, 120)
+        .then(result => {
+        expect(query.args[0][1]).to.deep.equal(expectedQueryParams);
```

Comment:
indentation is off here

Reply:
Ah this indentation thing gave me troubles this time. Fixed